### PR TITLE
Data view for Tiled Selection Dialog

### PR DESCRIPTION
**Pull request by AbbyGi**

### CODE REVIEW

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 1]
```diff
@@ -64,28 +64,28 @@ def __init__(
...
-        #       interacts with the URL text editor.
-        self.url_buffer = ""
+    @property
+    def url(self):
+        return self._url
+    
+    @url.setter
+    def url(self, value: str):
+        self._url = value
+        self._url_buffer = value
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]: I suppose we might want to call `self.url_changed.emit()` here.

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 2]
```diff
@@ -64,28 +64,28 @@ def __init__(
...
 
-        self.url_buffer = new_text
+        self._url_buffer = new_text
 
     def on_url_editing_finished(self):
         """Handle a notification that URL editing is complete."""
         _logger.debug("TiledCatalogSelector.on_url_editing_finished()...")
 
-        new_url = str.strip(self.url_buffer or "")
+        new_url = str.strip(self._url_buffer or "")
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]: Now that `self._url_buffer` will always be a string (and not `None`), we should be able to go back to this simpler version.

```suggestion
        new_url = self._url_buffer.strip()
```

File: PyMca5/PyMcaGui/io/QTiledCatalogSelectorDialog.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 from PyQt5.QtCore import QEvent, QObject
 from PyQt5.QtWidgets import (
     QDialog, QLabel, QLineEdit, QPushButton, QVBoxLayout, QWidget,
+    QApplication, QMainWindow,
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/QTiledCatalogSelectorDialog.py]: Let's move these to the testing section where they are used.

```suggestion
```

File: PyMca5/PyMcaGui/io/QTiledCatalogSelectorDialog.py [hunk 4]
```diff
@@ -116,15 +117,24 @@ def connect_model_slots(self) -> None:
...
         self.url_entry.editingFinished.connect(model.on_url_editing_finished)
         self.connect_button.clicked.connect(model.on_connect_clicked)
 
     def initialize_values(self) -> None:
         """Initialize widget values."""
         self.reset_url_entry()
+
+if __name__ == '__main__':
+    from sys import argv
+    app = QApplication(argv)
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/QTiledCatalogSelectorDialog.py]:
```suggestion
    from sys import argv
    from PyQt5.QtWidgets import QApplication, QMainWindow
    app = QApplication(argv)
```

**Review by padraic-shafer — COMMENTED** [review]
This is looking good.  I've made a few suggestions here.

I'll take a closer look at the tests and will probably submit a PR to your PR. ;-)

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 5]
```diff
@@ -25,50 +25,33 @@ def test_init(optional_args: Mapping[str, Any]):
     TiledCatalogSelector()
 
 
-def test_on_url_focus_in_event():
-    """Event handler creates a string buffer for a new url without changing the existing url."""
+def test_url_property():
+    """Model URL properly returned and set"""
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py]:
```suggestion
    """Model URL is properly returned and set."""
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 6]
```diff
@@ -25,50 +25,33 @@ def test_init(optional_args: Mapping[str, Any]):
...
-    event = Mock()
-
     assert model.url == expected_url
 
-    model.url_buffer = "Pre-existing content"
-    model.on_url_focus_in_event(event)
-
-    assert model.url_buffer == ""
+    expected_url = "after"
+    model.url = expected_url
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py:36]:
```suggestion
    expected_url = "after"
    model.url = expected_url
    # This is a non-trivial assertion when property getter/setter is used
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 7]
```diff
@@ -25,50 +25,33 @@ def test_init(optional_args: Mapping[str, Any]):
...
-    model.on_url_focus_in_event(event)
-    assert model.url_buffer == ""
-
-
 def test_on_url_text_edited():
     """Event handler replaces the url buffer without changing the existing url."""
     expected_url = "before"
     model = TiledCatalogSelector(url=expected_url)
-    model.url_buffer = ""
+    model._url_buffer = ""
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py]: This is now initialized during model initialization.

```suggestion
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 8]
```diff
@@ -25,50 +25,33 @@ def test_init(optional_args: Mapping[str, Any]):
...
 
     assert model.url == expected_url
-    assert model.url_buffer == ""
+    assert model._url_buffer == ""
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py]:
```suggestion
    # Buffer is pre-populated for the case when no user editing events have been received.
    assert model._url_buffer == expected_url
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 9]
```diff
@@ -77,14 +60,14 @@ def test_on_url_editing_finished():
     """Event handler replaces the existing url with the contents of the buffer."""
     expected_url = "after"
     model = TiledCatalogSelector(url="before")
-    model.url_buffer = expected_url
+    model._url_buffer = expected_url
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py:72]:
```suggestion
    # Behave as if the user had edited the url in the dialog widget
    model._url_buffer = expected_url
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 10]
```diff
@@ -192,7 +175,7 @@ def test_validation_on_url_editing_finished(
     caplog.set_level(logging.INFO)
     validators = {"url": [validate_url_scheme]}
     model = TiledCatalogSelector(url="before", validators=validators)
-    model.url_buffer = url
+    model._url_buffer = url
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py:190]:
```suggestion
    # Behave as if the user had edited the url in the dialog widget
    model._url_buffer = url
```

**Review by padraic-shafer — COMMENTED** [review]
Added a few more updates to the tests.

File: PyMca5/tests/ut/test_q_tiled_catalog_selector_dialog.py [hunk 11]
```diff
@@ -71,17 +71,23 @@ def test_url_editing(qtbot: QtBot, dialog_model: TiledCatalogSelector):
...
 
-
-def test_url_edit_focus(
-    qapp: QApplication,
-    qtbot: QtBot,
-    dialog_model: TiledCatalogSelector,
-):
-    """Verify custom event filtering for url_entry widget."""
+def test_url_set_through_model(qtbot: QtBot, dialog_model: TiledCatalogSelector):
+    """Verify setting the model url connects properly."""
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_q_tiled_catalog_selector_dialog.py]:
```suggestion
    """Connects to a url that is set on the model after dialog creation."""
```

File: PyMca5/tests/ut/test_q_tiled_catalog_selector_dialog.py [hunk 12]
```diff
@@ -71,17 +71,23 @@ def test_url_editing(qtbot: QtBot, dialog_model: TiledCatalogSelector):
...
+        client = Mock()
+        client.uri = dialog_model.url
+        client.context.api_uri = ""
+        mock_client_constructor.return_value = client
+
+        dialog.connect_button.click()
+
+    assert expected_url in dialog.connection_label.text()
+
+    assert expected_url in dialog.connection_label.text()
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_q_tiled_catalog_selector_dialog.py]: This assertion showed up twice.  Not sure if you meant to make a different assertion, or if it just got duplicated accidentally.

```suggestion
```

**Review by padraic-shafer — CHANGES_REQUESTED** [review]
I've now finished this round of reviewing.  Thanks again for keeping this progressing!

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 13]
```diff
@@ -84,10 +84,38 @@ def url(self, value: str):
...
         self._url = value
         self._url_buffer = value
-        # TODO: Perhaps this should only emit if the value is changed?
-        self.url_changed.emit()
+        if value != old_value:
+            self.url_changed.emit()
+
+    @property
+    def client(self):
+        return self._client
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]: Suggested docstring

```suggestion
    @property
    def client(self):
        """Fetch the root Tiled client."""
        return self._client
```

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 14]
```diff
@@ -84,10 +84,38 @@ def url(self, value: str):
...
+
+    @client.setter
+    def client(self, _):
+        raise NotImplementedError("Call set_root_client() instead")
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]: Suggested docstring

```suggestion
    @client.setter
    def client(self, _):
        """Do not directly replace the root Tiled client."""
        raise NotImplementedError("Call set_root_client() instead")
```

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 15]
```diff
@@ -84,10 +84,38 @@ def url(self, value: str):
...
+
+    def set_root_client(self) -> None:
+        """Update the model with content from the new root client.
+        
+            Emits the 'table_changed' signal when a client is defined."""
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]:
```suggestion
        """Update the model with content from the new root client.
        
            Emits the 'table_changed' signal when a client is defined.
            Emits the 'client_connection_error' signal when client does not connect."""
```

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 16]
```diff
@@ -84,10 +84,38 @@ def url(self, value: str):
...
+            _logger.error(error_message)
+            self.client_connection_error.emit(error_message)
+            return
+
+        self._client = new_client
+        self.client_connected.emit(self._client.uri, str(self._client.context.api_uri))
+        self.node_path_parts = ()
+        self._current_page = 0
+        if self.client is not None:
+            self.table_changed.emit(self.node_path_parts)
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py]: Adding some whitespace for logical separation.

```suggestion
        self._client = new_client
        self.client_connected.emit(self._client.uri, str(self._client.context.api_uri))

        self.node_path_parts = ()
        self._current_page = 0
        if self.client is not None:
            self.table_changed.emit(self.node_path_parts)
```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 17]
```diff
@@ -72,7 +78,7 @@ def test_on_url_editing_finished():
     assert model.url == expected_url
     assert model._url_buffer == expected_url
 
-    mock_signal.emit.assert_called_once_with()
+    getattr(mock_signal.emit, expected_emit)()
```
Comment (padraic-shafer) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py]: What do you think about making this a self-documenting assertion?

```suggestion
    assert_expected_emit = getattr(mock_signal.emit, expected_emit)
    assert_expected_emit()
```

**Review by padraic-shafer — CHANGES_REQUESTED** [review]
Nice!  Here are few things to clean up before merging.

**Comment by padraic-shafer** [comment]
@AbbyGi 
Following my latest review, I realized that `set_root_client()` is really doing 2 things.  Breaking them up into single-responsibility functions might look something like this...

```python3
    def on_connect_clicked(self, checked: bool = False):
        """Handle a button click to connect to the Tiled client."""
        _logger.debug("TiledCatalogSelector.on_connect_clicked()...")

        if self.client:
            # TODO: Clean-up previously connected client?
            ...

        self.connect_client()
        self.reset_client_view()

    def connect_client(self) -> None:
        """Connect the model's Tiled client to the Tiled server at URL.
        
            Emits the 'client_connection_error' signal when client does not connect.
        """
        try:
            new_client = self.client_from_url(self.url)
        except Exception as exception:
            error_message = str(exception)
            _logger.error(error_message)
            self.client_connection_error.emit(error_message)
            return

        self._client = new_client
        self.client_connected.emit(
            self._client.uri,
            str(self._client.context.api_uri),
        )

    def reset_client_view(self) -> None:
        """Prepare the model to receive content from a Tiled server.
        
            Emits the 'table_changed' signal when a client is defined.
        """
        self.node_path_parts = ()
        self._current_page = 0
        if self.client is not None:
            self.table_changed.emit(self.node_path_parts)

```

File: PyMca5/tests/ut/test_tiled_catalog_selector.py [hunk 18]
```diff
@@ -25,68 +25,61 @@ def test_init(optional_args: Mapping[str, Any]):
     TiledCatalogSelector()
```
Comment (AbbyGi) [inline PyMca5/tests/ut/test_tiled_catalog_selector.py]: @padraic-shafer should the `optional_args` be passed into the `TiledCatalogSelector` in the `test_init` test? Looks like they aren't being used right now.

Comment (padraic-shafer) [reply PyMca5/tests/ut/test_tiled_catalog_selector.py]: Great catch!  I missed that earlier.
```suggestion
    TiledCatalogSelector(**optional_args)
```

**Comment by padraic-shafer** [comment]
What do you think about splitting up `set_root_client()` into `connect_client ()` + `reset_client_view()`?...or something along those lines?
https://github.com/NSLS2/pymca/pull/17#issuecomment-2431943794

**Comment by AbbyGi** [comment]
I addressed splitting up `set_root_client` in 12aca0c50fae951101cc76361069bbd8aadd666d.

File: PyMca5/PyMcaGui/io/TiledCatalogSelector.py [hunk 19]
```diff
@@ -99,12 +99,12 @@ def client(self):
     def client(self, _):
         """Do not directly replace the root Tiled client."""
         raise NotImplementedError("Call set_root_client() instead")
-
-    def set_root_client(self) -> None:
-        """Update the model with content from the new root client.
+
```
Comment (padraic-shafer) [inline PyMca5/PyMcaGui/io/TiledCatalogSelector.py:102]:
```suggestion

```

**Review by padraic-shafer — APPROVED** [review]
Alright!

**Pull request merged**